pennylane/beta/tapes/tape.py

```diff
@@ -0,0 +1,777 @@
+        self._output_dim = 0
+
+        for obj, info in self._queue.items():
+
+            if isinstance(obj, QuantumTape):
+                self._ops.append(obj)
+
+            elif isinstance(obj, qml.operation.Operation) and not info.get("owner", False):
+                # operation objects with no owners
+                if self._obs:
+                    raise ValueError(
+                        f"Quantum operation {obj} must occur prior to any measurements."
+                    )
+
+                obj.do_check_domain = False
+
+                # invert the operation if required
+                obj.inverse = info.get("inverse", False)
+
+                if isinstance(obj, STATE_PREP_OPS):
+                    if self._ops:
+                        raise ValueError(
+                            f"State preparation operation {obj} must occur prior to any quantum operations."
+                        )
+
+                    self._prep.append(obj)
+                else:
+                    self._ops.append(obj)
+
+            elif isinstance(obj, MeasurementProcess):
+                # measurement process
+
+                if obj.return_type is qml.operation.Probability:
+                    self._obs.append((obj, obj))
+                    self._output_dim += 2 ** len(obj.wires)
+
+                elif "owns" in info:
+                    # TODO: remove the following line once devices
+                    # have been refactored to no longer use obs.return_type.
+                    # Monkeypatch the observable to have the same return
+                    # type as the measurement process.
+                    info["owns"].return_type = obj.return_type
+                    info["owns"].do_check_domain = False
+
+                    self._obs.append((obj, info["owns"]))
+                    self._output_dim += 1
+
+                    if obj.return_type is qml.operation.Sample:
+                        self.is_sampled = True
+
+            elif isinstance(obj, qml.operation.Observable) and "owner" not in info:
+                raise ValueError(f"Observable {obj} does not have a measurement type specified.")
+
+        self._update()
+
+    def _update_circuit_info(self):
+        """Update circuit metadata"""
+        self.wires = qml.wires.Wires.all_wires(
+            [op.wires for op in self.operations + self.observables]
+        )
+        self.num_wires = len(self.wires)
+
+    def _update_par_info(self):
+        """Update the parameter information dictionary"""
+        param_count = 0
+
+        for obj in self.operations + self.observables:
+
+            for p in range(len(obj.data)):
+                info = self._par_info.get(param_count, {})
+                info.update({"op": obj, "p_idx": p})
+
+                self._par_info[param_count] = info
+                param_count += 1
+
+    def _update_trainable_params(self):
+        """Set the trainable parameters"""
+        self._trainable_params = set(self._par_info)
```

Comment:
Is this the one that gets overwritten by the interfaces?

Reply:
yep!